src/main/java/org/terracotta/offheapstore/paging/UpfrontAllocatingPageSource.java

```diff
@@ -440,67 +473,146 @@ public synchronized Runnable removeAllocationThreshold(ThresholdDirection direct
-        ByteBuffer b = source.allocateBuffer((int) Math.min(currentChunkSize, (max - allocated)));
-        long blockDuration = System.nanoTime() - blockStart;
-        if (b == null) {
-          if (fixed || (currentChunkSize >>> 1) < minChunk) {
-            throw new IllegalArgumentException("An attempt was made to allocate more off-heap memory than the JVM can allow." +
-                    " The limit on off-heap memory size is given by the -XX:MaxDirectMemorySize command (or equivalent).");
-          } else {
-            LOGGER.debug("Allocated {}B in {}B chunks.", new Object[]{DebuggingUtils.toBase2SuffixedString(allocated - lastFailureAt), DebuggingUtils.toBase2SuffixedString(currentChunkSize)});
-            currentChunkSize >>>= 1;
-            lastFailureAt = allocated;
-          }
-        } else {
-          buffers.add(b);
-          allocated += b.capacity();
-          
-          if (allocatorLog != null) {
-            allocatorLog.printf("%d,%d,%d,%d,%d,%d,%d,%d%n", System.nanoTime() - start, blockDuration, b.capacity(), allocated, PhysicalMemory.freePhysicalMemory(), PhysicalMemory.totalSwapSpace(), PhysicalMemory.freeSwapSpace(), PhysicalMemory.ourCommittedVirtualMemory());
-          }
-          if (allocated > nextProgressLogAt) {
-            LOGGER.info("Allocation {}% complete", (100 * allocated) / max);
-            nextProgressLogAt += progressStep;
+        if (allocatorLog != null) {
+          allocatorLog.printf("timestamp,duration,size,physfree,totalswap,freeswap,committed%n");
+        }
+
+        ExecutorService executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
+
+        List<Future<Long>> futures = new ArrayList<Future<Long>>((int)toAllocate / maxChunk + 1);
+
+        for (long dispatched = 0; dispatched < toAllocate; ) {
+          final int currentChunkSize = (int)Math.min(maxChunk, toAllocate - dispatched);
+          futures.add(executorService.submit(new Callable<Long>() {
+            @Override
+            public Long call() throws Exception {
+              return bufferAllocation(source, currentChunkSize, minChunk, fixed, allocatorLog, start, buffers);
+            }
+          }));
+          dispatched += currentChunkSize;
+        }
+
+        executorService.shutdown();
+
+        long allocated = 0;
+        long progressStep = Math.max(PROGRESS_LOGGING_THRESHOLD, (long)(toAllocate * PROGRESS_LOGGING_STEP_SIZE));
+        long nextProgressLogAt = progressStep;
+
+        for (Future<Long> future : futures) {
+          try {
+            allocated += future.get();
+            if (allocated > nextProgressLogAt) {
+              LOGGER.info("Allocation {}% complete", (100 * allocated) / toAllocate);
+              nextProgressLogAt += progressStep;
+            }
+          } catch (ExecutionException e) {
+            if (e.getCause() instanceof RuntimeException) {
+              throw (RuntimeException)e.getCause();
```

Comment:
We could, but since things are going wrongs already, I'm not sure we won't add to the mess. We could add a shutdownNow. That's what you had in mind?